src/gensxg.c

```diff
@@ -0,0 +1,444 @@
+  const char* content_type;
+  sxg_header_t header;
+  uint64_t mi_record_size;
+  const char* url;
+  const char* cert_url;
+  const char* validity_url;
+  const char* certificate;
+  const char* public_key;
+  const char* private_key;
+  time_t date;
+  int64_t duration;
+} Options;
+
+static EVP_PKEY* load_private_key(const char* filepath) {
+  FILE* const keyfile = fopen(filepath, "r");
+  if (keyfile == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
+  }
+  EVP_PKEY* private_key =
+      PEM_read_PrivateKey(keyfile, NULL, NULL, NULL);
+  if (private_key == NULL) {
+    fprintf(stderr, "Failed to read private key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
+  fclose(keyfile);
+  return private_key;
+}
+
+static X509* load_x509_cert(const char* filepath) {
+  FILE* certfile = fopen(filepath, "r");
+  if (certfile == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
+  }
+  X509* cert = PEM_read_X509(certfile, NULL, NULL, NULL);
+  if (cert == NULL) {
+    fprintf(stderr, "Failed to read cetificate: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
+  fclose(certfile);
+  return cert;
+}
+
+static EVP_PKEY* load_ed25519_pubkey(const char* filepath) {
+  FILE* keyfile = fopen(filepath, "r");
+  if (keyfile == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
+  }
+  EVP_PKEY* public_key = PEM_read_PUBKEY(keyfile, NULL, NULL, NULL);
+  if (public_key == NULL) {
+    fprintf(stderr, "Failed to read public key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
+  fclose(keyfile);
+  return public_key;
+}
+
+static time_t parse_time(const char* datestring) {
+  struct tm date;
+  char* finished = strptime(datestring, "%Y-%m-%dT%H:%M:%SZ", &date);
+  if (finished == NULL) {
+    fprintf(stderr, "Failed to parse date: %s\n", datestring);
+  } else if(*finished != '\0') {
+    fprintf(stderr, "Failed to parse at: %s\n", finished);
+    exit(EXIT_FAILURE);
+  }
+  return mktime(&date);
+}
+
+static Options init_default_options() {
+  Options result;
+  memset(&result, 0, sizeof(result));
+  result.content = "index.html";
+  result.content_type = "text/html";
+  result.header = sxg_empty_header();
+  result.date = time(NULL); /* = Now. */
+  result.duration = 60 * 60 * 24 * 7;  /* = 7 Days. */
+  result.mi_record_size = 4096;
+  return result;
+}
+
+static void header_append_string(const char* data, sxg_header_t* header) {
+  size_t colon_pos = 0;
+  for (size_t i = 0; data[i] != '\0'; ++i) {
+    if (data[i] == ':') {
+      colon_pos = i;
+      break;
+    }
+  }
+  if (colon_pos == 0) {
+    fprintf(stderr, "Invalid header format: %s\n", data);
+    exit(EXIT_FAILURE);
+  }
+  size_t value_pos = colon_pos + 1;
+  while (data[value_pos] != '\0' && data[value_pos] == ' ') ++value_pos;
+  char* const key = OPENSSL_strndup(data, colon_pos);
+  char* const value = OPENSSL_strdup(&data[value_pos]);
```

Comment:
* The trailing space needs to be truncated, to be strict, though I'm not sure how much it matters.
* Optional: `data + value_pos` looks more natural to me (but it's probably a matter of taste).

Reply:
Done.